src/lib.rs

```diff
@@ -3,27 +3,1109 @@
+pub struct StakedState<BlockNumber, Balance, SlashId: Ord> {
+    /// Total amount of funds at stake.
+    pub staked_amount: Balance,
+
+    /// Status of the staking.
+    pub staked_status: StakedStatus<BlockNumber>,
+
+    /// SlashId to use for next Slash that is initiated.
+    /// Will be incremented by one after adding a new Slash.
+    pub next_slash_id: SlashId,
+
+    /// All ongoing slashing.
+    pub ongoing_slashes: BTreeMap<SlashId, Slash<BlockNumber, Balance>>,
+}
+
+impl<BlockNumber, Balance, SlashId> StakedState<BlockNumber, Balance, SlashId>
+where
+    BlockNumber: SimpleArithmetic + Copy,
+    Balance: SimpleArithmetic + Copy,
+    SlashId: Ord + Copy,
+{
+    /// Iterates over all ongoing slashes and decrements blocks_remaining_in_active_period_for_slashing of active slashes (advancing the timer).
+    /// Returns the the number of slashes that were found to be active and had their timer advanced.
+    fn advance_slashing_timer(&mut self) -> usize {
+        self.ongoing_slashes
+            .iter_mut()
+            .fold(0, |slashes_updated, (_slash_id, slash)| {
+                if slash.is_active
+                    && slash.blocks_remaining_in_active_period_for_slashing > Zero::zero()
+                {
+                    slash.blocks_remaining_in_active_period_for_slashing -= One::one();
+                    slashes_updated + 1
+                } else {
+                    slashes_updated
+                }
+            })
+    }
+
+    /// Returns pair of slash_id and slashes that should be executed
+    fn get_slashes_to_finalize(&mut self) -> Vec<(SlashId, Slash<BlockNumber, Balance>)> {
+        let slashes_to_finalize = self
+            .ongoing_slashes
+            .iter()
+            .filter(|(_, slash)| {
+                slash.blocks_remaining_in_active_period_for_slashing == Zero::zero()
+            })
+            .map(|(slash_id, _)| *slash_id)
+            .collect::<Vec<_>>();
+
+        // remove and return the slashes
+        slashes_to_finalize
+            .iter()
+            .map(|slash_id| {
+                // assert!(self.ongoing_slashes.contains_key(slash_id))
+                (*slash_id, self.ongoing_slashes.remove(slash_id).unwrap())
+            })
+            .collect()
+    }
+
+    /// Executes a Slash. If remaining at stake drops below the minimum_balance, it will slash the entire staked amount.
+    /// Returns the actual slashed amount.
+    fn apply_slash(
+        &mut self,
+        slash: Slash<BlockNumber, Balance>,
+        minimum_balance: Balance,
+    ) -> Balance {
+        // calculate how much to slash
+        let mut slash_amount = if slash.slash_amount > self.staked_amount {
+            self.staked_amount
+        } else {
+            slash.slash_amount
+        };
+
+        // apply the slashing
+        self.staked_amount -= slash_amount;
+
+        // don't leave less than minimum_balance at stake
+        if self.staked_amount < minimum_balance {
```

Comment:
Why do we do this? What would be the effect of just dipping below the minimum of `Currency` and not doing this?

Reply:
This is to ensure that the sum of all staked_amount values of all Stakes will always be equal to the value in the stake pool. Otherwise we could get a situation where the stake pool has zero balance (because it is an actual account and will follow the existential deposit rules) but a stake would have a positive value. This is problematic if an attempt is then made to unstake that amount which would not be available in the pool.